### ISSUE TITLE
Implement a widening operation

### ISSUE THREAD

**Issue by eddywestbrook** [GaloisInc/heapster-saw#84]
Currently, users have to manually write type-checking hints for loops that iterate over data structures. If we wrote a widening operation, much of this could be done automatically, greatly reducing the user effort required to type-check C functions with Heapster.

The high-level design is to have the type-checker notice when there is a backwards jump whose required permissions cannot be satisfied (i.e., when the implication fails), and to then generalize those required permissions by "widening" them against the current permissions. Widening here is a binary operation on permissions that is similar to taking the "or" of two permissions, in that `widen p1 p2` is a lower bound for both `p1` and `p2`, but instead of taking the greatest lower bound like or does, it tries to take a lower bound that is more likely to be satisfied by the next iteration of the loop.

The common example will be if we widen permissions `x:eq(y), y:ListR<e>` against `x:ptr(...) * ptr((W,8) |-> eq(y)), y:ListR<e>` where `ListR` is the reachability permission for linked lists. Intuitively this means that each iteration of the loop is probably going to follow the `next` pointer, so we want to generalize to permissions `x:ListR<y>, y:ListR<e>`. Some pairs of permissions will not make sense to widen with each other, and so `widen` will be a partial operation.

This feature should require 3 steps:
1. Modify the type-checking loop for CFGs to be able to revisit nodes that have already been type-checked when we widen them;
2. Modify `tcJumpTarget` to notice when backwards jumps do not succeed and insert a new entrypoint for the widened input permissions; and
3. Implement the binary `widen` operation

**Comment by eddywestbrook** [comment]
As a first step, we have to separate out the computation of in-degrees of blocks so that it happens before any calls to `visit` (which performs type-checking), because we are now potentially going to type-check the same block multiple times and we don't want to have those multiple visits keep incrementing the in-degree.

**Comment by eddywestbrook** [comment]
Specifically, the in-degree should be moved up to `TypedBlock` instead of having it be per-entry, and it should either be `Single` or `Multi`. (Let's keep it a separate type rather than just a `Bool` so its semantics are more clear.) The logic is:

* If a block is the head of a loop, its in-degree is always `Multi`
* If a block has actual (in the graph sense) in-degree >= 2 and is a join point, its in-degree is `Multi`
* Otherwise the in-degree is `Single`

The logic behind the last case is that, even if there are multiple jumps to a block, they will all become separate entry points.

**Comment by eddywestbrook** [comment]
Issue: when we widen the head of a loop, we need to move backwards in our topologically sorted list of blocks to visit. This somehow needs to be communicated from `tcJumpTarget` inside the `PermCheckM` parameterized monad up to the top-level loop over the blocks in `tcCFG`.

IDEA: we could change `liftInnerToTopM` to have a second return value that contains any `BlockID`s for blocks that have been marked as needing to be re-visited.

BETTER: we could just change `TopPermCheckState` to contain a list of blocks that have been marked as needing to be re-visited, and `applyDeltasToTopState` could stash the `BlockID`s in there, using a new constructor for `BlockInfoMapDelta`.

**Comment by eddywestbrook** [comment]
ISSUE: When we widen a block, we also need to re-visit all entrypoints it calls. To do that, we need to delete or invalidate the old versions of those entrypoints, so we can calculate their new input permissions. If any of these are join points that are not dominated by the block we are widening, then we need to fix up all the callers of these join points to call the new entrypoints with their new permissions. There is currently no way to do this fixing up.

IDEA: We should have two sorts of blocks, "join" blocks and "multi-entry" blocks. Multi-entry blocks have one entrypoint per caller. These entrypoints are effectively the same as multiple copies of the same block, and have nothing to do with each other. Join blocks, on the other hand, have one entrypoint, and all callers get funneled into calls to that one entrypoint. The *cool new thing* that is going to address our issue above is that join blocks are themselves going to maintain a list of their callers, by `TypedEntryID`, along with the permissions held by each caller at its call site and the `PermImpl` required to prove the entrypoint input permissions from those permissions. This allows us to modify the permissions of just one entrypoint corresponding to a particular caller, so that when we are re-visiting blocks we can update the permissions of just that one call. It also means that we can wait on generating the actual entrypoint input permissions until we have seen all the callers, and can even modify those input permissions without modifying the callers.

NOTE: the notion of join vs multi-entry blocks should also subsume the in-degree functionality, whose sole purpose was to determine which entrypoints should become letrec-bound functions and which just have their translations folded into that of their callers. Now we can have the join blocks be the former and the multi-entry blocks be the latter. Note that this also means we want to ensure join blocks do have multiple callers, otherwise we are making functions be letrec-bound without needing to. Though OTOH we can probably convert join blocks with only one caller to multi-entry blocks.

**Comment by eddywestbrook** [comment]
ISSUE: widening can change the ghost variables for an entrypoint, so entrypoint IDs should not contain ghost variables.

IDEA: In fact, maybe we should make a separate `TypedEntryBody` type, that includes the ghost variables, input permissions, and all the implications from call sites to the entry point body

**Comment by eddywestbrook** [comment]
Another idea: instead of jumping backwards in our topological order when we have a back edge, we should just do multiple passes through the entire list of blocks. This will be much easier to implement, and has the added benefit that it is much easier to figure out the logic for preventing infinite widening loops: just limit the number of iterations through the list of blocks. In the old way, we could end up re-checking an inner loop multiple times before re-checking the outer loop containing it, which would in general require quadratically many visits to the innermost loop in a function.

**Comment by eddywestbrook** [comment]
The actual widening operation tries to combine two sets of permissions on the same arguments but with different evars, where the evars could also have permissions. That is, we want to combine an `MbValuePerms (args :++: vars1)` with an `MbValuePerms (args :++: vars2)`. The main criterion is that, if the evars in each input set are determined by the permissions on the `args`, then the evars in the output set should be determined by the output permissions on the `args`. Recall that evars are determined by being used as a read/write or lifetime modality, by being in an equality permission in the form `eq(x &+ off)`, or by being in a pointer, array, or block permission whose contents determine them. Thus, for instance, existential and disjunctive permissions do not determine any evars.

IDEA: One approach to our algorithm would be to essentially make a unification algorithm, where we traverse the permissions on `args`, making sure that both sides mostly use the same constructors. For the case `widen (eq(x)) p` for `x` unset, we make a new output evar `y`, set `x := y` in the LHS evars, lookup the LHS perms `p_x` on `x` on the LHS, and recursively call `widen p_x p` to get the output permissions on `y`. If `x` is already set to some expression `e`, then we could go two ways: we could require `p` to be an equality permission `eq(e')` such that `e'` unifies with `e`; or we could, if `e=y`, take the existing permissions on `y` and split them between `y` and the current permissions we are solving for, by examining `p`. The former would be simpler, and in fact is probably all we need, because the latter is effectively like splitting up the fields of a C struct type between two different pointers, which is rarely what is needed.

The difference from normal unification is that a fresh output evar is the most general expression, because it can be instantiated in any way. Another way to look at it is that a fresh output evar represents the disjunction over all possible expressions at a particular location, which is the least upper bound of each individual expression that could be used there. Note, however, that we cannot introduce fresh output evars at locations that do not determine their values, because then we would violate our main criterion from above.

One remaining question is how to handle existential permissions. At the top level, an existential permission `x:exists z.p` on either side can be handled by making a fresh output evar `y` for `z` and then widening the other size against `[y/z]p`. The issue comes when `p` is in a location that does not determine any evars; e.g., if we have `p1 or (exists z.p2)`.

IDEA: Can we locally capture all the output evars created in a computation? That is, if we widen `p1` against `p2` to get `p` with output evars `z1, ..., zn` created by that widening, can we capture the `zi` and return `exists z1, ..., zn. p`? If so, this definitely seems like the way to go, because then those variables only need to be determined in that local context. For example, if we are widening `p1 or p2` against `p1 or p2'` and widening `p2` against `p2'` yields some `p2''` that determines some evars `z1, .., zn`, then it is not ok to return `p1 or p2''` but it _is_ ok to return `p1 or (exists z1,...,zn. p2'')` because the `zi` do not have to be determined unless the implication prover chooses the right-hand disjunct, in which case `p2''` becomes top-level and they are determined.

ISSUE: but this does not work because local evars cannot have permissions on them, whereas global ones can.

**Comment by eddywestbrook** [comment]
IDEA: disjunctive and existential permissions are not generally created by type-checking or implication-proving, but instead usually only come from user annotations, either on function inputs or outputs or by unfolding a named permission. So these probably correspond to whole types in the original program, rather than fragments split up and borrowed by different variables, and so it makes sense to just do unification on them, rather than trying to widen them against each other.

Specifically, the widening unification algorithm, `widUnifyM`, should have type

`Mb (vars1 :++: vars) a -> Mb (vars2 :++: vars) a -> Mb vars a`

to allow unification under existential permissions that bind the same variables `vars`. If this unification ever comes across a variable in `vars1` or `vars2`, both sides should become a fresh output evar, and normal widening should happen for the perms for the two sides.

**Comment by eddywestbrook** [comment]
IDEA: instead of having a separate unification-y algorithm on permissions in binders, why don't we make our general widening algorithm work on permissions in binders? If we get rid of the idea of the idea of turning existential permissions into fresh output evars, then this will work inside disjunctions, array permissions, and other places where it doesn't make sense to generate fresh output evars. If we do come across a top-level evar on one or both sides, and so need to generate an output evar, we widen its permissions against an existential permission that binds all the local existential vars on the other side.

**Comment by eddywestbrook** [comment]
IDEA: instead of ever having widening fail, we can always just return `true` permissions. The idea of widening is that it must return a lower bound, but we are trying heuristically to make it a somewhat tight lower bound, modulo our criterion above and modulo trying to make it loose enough so that it is inductive for loops.

**Comment by eddywestbrook** [comment]
IDEA: there are times when it might make sense to split permissions during widening. That is, consider widening

`x:p, y:eq(x)`   against   `x:p1, y:p2`

This requires somehow splitting `p` into permissions that can be widened against `p1` and `p2`. As mentioned above, it doesn't really make sense to split `p` into fields on different offsets, as this is like splitting a C struct into different fields held by different points. But if, for instance, `p` is copyable, then permission splitting could just copy `p` across `x` and `y`. As another possibility, we could also return permissions like `x:p1, y:(eq(x) or p2)`, as long as `p2` isn't needed to determine any variables.

FUTURE WORK: this could be useful in the future, but let's get a version of widening finished before trying to add this...

**Comment by eddywestbrook** [comment]
IDEA: maybe this would be so much easier if we start out by binding fresh output evars for all the evars in both the LHS and RHS, and then iteratively unify some of them as part of widening? The benefit would be that we would not have to work with multi-bindings in different evar contexts, which would simplify things.

ISSUE: the straightforward way to widen `eq(x)` against `p` is to leave the `eq(x)` perm and widen `x`'s perms against `p`, but this would lead to equating two locations in permission sets that are equated by either side, whereas we want to only equate those locations that are equated by both sides.

IDEA: what if we track what evars have been visited? An unvisited perm has no location in the top-level perm set yet, while a visited one does. Then, when we widen `eq(x)` against `p`, if `x` has already been visited, we split its existing perms using `p`, as discussed above. Only if `x` has not yet been visited do we put in an `eq(x)` permission.

**Comment by eddywestbrook** [comment]
ISSUE: how do we widen permissions in contexts where they can be repeated, such as in array permissions or arguments of recursively-defined permissions?

The key difficulty is expressions with locally-bound existential variables. For instance, if we widen

`array(..., [(W,0) |-> eq(llvmword(1))])` against `array(..., [(W,0) |-> exists x. eq(llvmword(8*x))])`

we should in theory get `exists x. eq(llvmword(x))`, whereas widening these two permissions against each other outside a repeated context should give `eq(llvmword(x))` with a globally-bound evar `x`. The different is that the former allows a different `x` in each array call, whereas the latter requires the same `x` for all of them.

IDEA: for now, we just require array permissions to be equal, or we return `true`. We don't really support implications inside array permissions now anyway. Same with arguments to recursively-defined permissions.

NOTE: this means we do _not_ have to handle locally-bound variables in our widening. That is, when we see `exists x.p`, we instantiate `x` with a fresh, global output evar.

IDEA FOR LATER: write a specialized `widenLocalPerm` that matches on permissions in bindings and returns a permission in an existential binding, i.e., a `SomeMb (ValuePerm a)`
